src/languages/c.coffee

```diff
@@ -3,42 +3,256 @@
+  ['selectionStatement', 'control'],
+  ['iterationStatement', 'control'],
+  ['functionDefinition', 'control'],
+  ['expressionStatement', 'command'],
+  ['expression', 'value'],
+  ['parameterDeclaration', 'command'],
+  ['unaryExpression', 'value'],
+  ['typeName', 'value'],
+  ['initializer', 'value'],
+  ['castExpression', 'value'],
+]
+
+SHAPE_RULES = [
+  ['blockItem', 'block-only'],
+  ['expression', 'value-only'],
+  ['postfixExpression', 'block-only'],
+  ['equalityExpression', 'value-only'],
+  ['logicalAndExpression', 'value-only'],
+  ['logicalOrExpression', 'value-only'],
+  ['iterationStatement', 'block-only'],
+  ['selectionStatement', 'block-only'],
+  ['assignmentExpression', 'block-only'],
+  ['relationalExpression', 'value-only'],
+  ['initDeclarator', 'block-only'],
+  ['externalDeclaration', 'block-only'],
+  ['structDeclaration', 'block-only'],
+  ['declarationSpecifier', 'block-only'],
+  ['statement', 'block-only'],
+  ['selectionStatement', 'block-only'],
+  ['iterationStatement', 'block-only'],
+  ['functionDefinition', 'block-only'],
+  ['expressionStatement', 'value-only'],
+  ['expression', 'value-only'],
+  ['additiveExpression', 'value-only'],
+  ['multiplicativeExpression', 'value-only'],
+  ['declaration', 'block-only'],
+  ['parameterDeclaration', 'block-only'],
+  ['unaryExpression', 'value-only'],
+  ['typeName', 'value-only'],
+  ['initializer', 'value-only'],
+  ['castExpression', 'value-only']
+]
+
 config = {
-  INDENTS, SKIPS, PARENS, SOCKET_TOKENS, COLORS_FORWARD, COLORS_BACKWARD
+  RULES, COLOR_RULES, SHAPE_RULES
 }
 
+ADD_PARENS = (leading, trailing, node, context) ->
+  leading '(' + leading()
+  trailing trailing() + ')'
+
+config.PAREN_RULES = {
+  'primaryExpression': {
+    'expression': ADD_PARENS
+    'additiveExpression': ADD_PARENS
+    'multiplicativeExpression': ADD_PARENS
+    'assignmentExpression': ADD_PARENS
+    'postfixExpression': ADD_PARENS
+  }
+}
+
+config.SHOULD_SOCKET = (opts, node) ->
+  unless opts.knownFunctions? and ((node.parent? and node.parent.parent? and node.parent.parent.parent?) or
+      node.parent?.type is 'specialMethodCall')
+    return true
+
+  # If it is a function call, and we are the first child
+  if (node.parent.type is 'primaryExpression' and
+     node.parent.parent.type is 'postfixExpression' and
+     node.parent.parent.parent.type is 'postfixExpression' and
+     node.parent.parent.parent.children.length in [3, 4] and
+     node.parent.parent.parent.children[1].type is 'LeftParen' and
+     (node.parent.parent.parent.children[2].type is 'RightParen' or node.parent.parent.parent.children[3]?.type is 'RightParen') and
+     node.parent.parent is node.parent.parent.parent.children[0] or
+     node.parent.type is 'specialMethodCall') and
+     node.data.text of opts.knownFunctions
+    return false
+  return true
+
+config.COLOR_CALLBACK = (opts, node) ->
+  return null unless opts.knownFunctions?
+
+  if node.type is 'postfixExpression' and
```

Comment:
This has some similarities to the expression just above.  Also needs to be explained. And can anything be factored into a common test function?